gonzo/clouds/compute.py

```diff
@@ -0,0 +1,254 @@
+from libcloud.compute.providers import get_driver as get_compute_driver
+
+
+logger = logging.getLogger(__name__)
+backends = {}
+
+
+def backend_for(provider):
+    def wrapper(cls):
+        backends[provider] = cls
+        return cls
+    return wrapper
+
+
+class Cloud(object):
+    compute_session = None
+
+    @classmethod
+    def from_config(cls, cloud_config, region):
+        try:
+            backend = cloud_config['BACKEND']
+        except KeyError:
+            raise LookupError("No backend specified!")
+
+        try:
+            backend_cls = backends[backend]
+        except KeyError:
+            raise LookupError(
+                "Unknown backend `{}`. Please choose one of {}".format(
+                    backend, backends.keys()))
+        return backend_cls(cloud_config, region)
+
+    def list_instances(self):
+        instances = self.compute_session.list_nodes()
+        for instance in instances:
+            self._monkeypatch_instance(instance)
+        return instances
+
+    def get_instance_by_uuid(self, instance_uuid):
+        for instance in self.list_instances():
+            if instance.uuid == instance_uuid:
+                return instance
+
+    def get_instance_by_name(self, instance_name):
+        for instance in self.list_instances():
+            if instance.name == instance_name:
+                return instance
+
+    def list_instances_by_type(self, instance_type):
+        all_instances = self.list_instances()
+        instance_types = []
+
+        for instance in all_instances:
+            metadata = self.compute_session.ex_get_metadata_for_node(
+                    instance)
+
+            if metadata.get('server_type') == instance_type:
+                instance_types.append(instance)
+
+        instance_types.sort(key=lambda i: i.name)
+        return instance_types
+
+    def list_instance_tags(self, node):
+        return node.extra[self.TAG_KEY]
+
+    def list_availability_zones(self):
+        return self.compute_session.list_locations()
+
+    def get_next_az(self, server_type):
+        availabie_azs = self.list_availability_zones()
+        try:
+            newest_instance_az = self.list_instances_by_type(
+                server_type)[-1].extra['gonzo_az']
+        except IndexError:
+            print "No Server Types Found"
+            return availabie_azs[0]
+
+        if len(availabie_azs) == 1:
+            return availabie_azs[0]
+        else:
+            for index, availabie_az in enumerate(availabie_azs):
+                if availabie_az.name == newest_instance_az:
+                    if (index + 1) == len(availabie_azs):
+                        return availabie_azs[0]
+                    else:
+                        return availabie_azs[index + 1]
+
+    def create_instance(self, image_name, name, owner, user_data=None,
+                        security_groups=None, size=None, key_name=None):
+        instance_name = name.split('-')
```

Comment:
actually, you might want something like 
`environment, server_type = name.split("-", 1)`